### PR TITLE
Track generated functions for torch compile

### PR DESCRIPTION
**Pull request by Ch0ronomato**
## Description
The torch compiler was having an issue running pytensor graphs with subgraphs. The way pytorch / module resolution worked was causing pytorch guards to fail during creation. Torch requires the guards to be created, and for them to be correct at runtime, so not being able to make one meant things couldn't get compiled. This tries to bandaid that by putting the modules we generate onto a module, essentially giving an explicit lifecycle to that module. This means we don't need to have graph breaks. That means faster compiler times, and better compiler results.

for this code
```py
import numpy as np
import torch._dynamo as dynamo

from pytensor.configdefaults import config
from pytensor.compile.builders import OpFromGraph
from pytensor.compile.function import function
from pytensor.tensor.type import matrices

x, y, z = matrices("xyz")
ofg_1 = OpFromGraph([x, y], [x + y])
ofg_2 = OpFromGraph([x, y], [x * y, x - y])

o1, o2 = ofg_2(y, z)
out = ofg_1(x, o1) / o2

xv = np.ones((2, 2), dtype=config.floatX)
yv = np.ones((2, 2), dtype=config.floatX) * 3
zv = np.ones((2, 2), dtype=config.floatX) * 5

f = function([x, y, z], out, mode="PYTORCH")
print(f(xv, yv, zv))

```

with change
```txt
Graph Count: 1
Graph Break Count: 0
Op Count: 4
Break Reasons:
Ops per Graph:
  Ops 1:
    <built-in method multiply of type object at 0x128d5a9a0>
    <built-in method subtract of type object at 0x128d5a9a0>
    <built-in method add of type object at 0x128d5a9a0>
    <built-in method true_divide of type object at 0x128d5a9a0>
Compile Times: TorchDynamo compilation metrics:
Function, Runtimes (s)
Number of Out Guards: 58
_compile.<locals>.compile_inner, 0.0773
OutputGraph.call_user_compiler, 0.0006
```

without change
```txt
(pytensor-dev) ➜  pytensor git:(33a4d4882) ✗ python test.py
Graph Count: 4
Graph Break Count: 3
Op Count: 4
Break Reasons:
Ops per Graph:
  Ops 1:
    <built-in method multiply of type object at 0x11ed5a9a0>
  Ops 2:
    <built-in method subtract of type object at 0x11ed5a9a0>
  Ops 3:
    <built-in method add of type object at 0x11ed5a9a0>
  Ops 4:
    <built-in method true_divide of type object at 0x11ed5a9a0>
Number of Out Guards: 140
Compile Times: TorchDynamo compilation metrics:
Function, Runtimes (s)
_compile.<locals>.compile_inner, 0.0397, 0.0174, 0.0145, 0.0056, 0.0147, 0.0151
OutputGraph.call_user_compiler, 0.0007, 0.0000, 0.0000, 0.0000
```

## Related Issue

- [ ] Closes #
- [ ] Related to #1076 

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1094.org.readthedocs.build/en/1094/

### CODE REVIEW

File: pytensor/link/pytorch/dispatch/basic.py [hunk 1]
```diff
@@ -56,9 +58,28 @@ def pytorch_funcify_FunctionGraph(
...
 ):
+    def conversion_func_register(*args, **kwargs):
+        last_frame = inspect.currentframe().f_back
+        functor = pytorch_funcify(*args, **kwargs)
+        if last_frame:
+            module = sys.modules[last_frame.f_globals["__name__"]]
+            name = functor.__name__
+            i = 0
+            prefix = ""
+            while True:
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/dispatch/basic.py:69]: This is just placeholder. We need to essentially make sure `elemwise_fn` and `elemwise_fn_1` show up in the guard.

File: pytensor/link/pytorch/dispatch/basic.py [hunk 2]
```diff
@@ -56,9 +58,17 @@ def pytorch_funcify_FunctionGraph(
     fgraph_name="pytorch_funcified_fgraph",
     **kwargs,
 ):
+    def conversion_func_register(*args, **kwargs):
+        last_frame = inspect.currentframe().f_back
+        functor = pytorch_funcify(*args, **kwargs)
+        if last_frame:
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/dispatch/basic.py:64]: It might be cleaner to just import `pytensor.link.utils`

File: tests/link/pytorch/test_basic.py [hunk 3]
```diff
@@ -335,7 +335,7 @@ def test_pytorch_OpFromGraph():
     ofg_2 = OpFromGraph([x, y], [x * y, x - y])
 
     o1, o2 = ofg_2(y, z)
-    out = ofg_1(x, o1) + o2
+    out = ofg_1(x, o1) / o2
```
Comment (Ch0ronomato) [inline tests/link/pytorch/test_basic.py:338]: This probably could be it's own commit, but c'est la vie.

File: pytensor/link/pytorch/dispatch/basic.py (hunk 2 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py]: I did this.

File: pytensor/link/pytorch/dispatch/basic.py [hunk 4]
```diff
@@ -56,9 +57,18 @@ def pytorch_funcify_FunctionGraph(
     fgraph_name="pytorch_funcified_fgraph",
     **kwargs,
 ):
+    # Ensure that torch is aware of the leaf nodes
+    # of generated code so we can compile code
+    # without graph breaks
+    def conversion_func_register(*args, **kwargs):
+        functor = pytorch_funcify(*args, **kwargs)
+        module = pytensor.link.utils
+        setattr(module, kwargs["unique_name"](functor), functor)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:66]: Do we need a weakref perhaps? So memory does get freed if nothing else is using these functions?

File: tests/link/pytorch/test_basic.py (hunk 3 above)
Comment (Ch0ronomato) [reply tests/link/pytorch/test_basic.py:338]: Torch makes an individual guard for each `Elemwise._check_runtime_shapes`. So we would have a guard for multiplication, subtraction, addition and division. I just wanted to keep all of them unique so I could let stuff blow up appropriately.

File: pytensor/link/pytorch/dispatch/basic.py (hunk 4 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: I tried doing weakref (ala `weakref.ref`) in both the leaf node (elemwise dispatch) and in the conversion function (where I current register the module). It seems like link/utils.py has some functionality that takes over the ownership of result - so we would have to make a change to that section. I can give that a shot, but it feels more intrusive (all backends would have weakrefs)

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/basic.py:66]: The problem with this is PyTensor becomes a giant memory leak if you compile enough pytensor functions?

Can we get input from the torch devs now that we narrowed down the problem?

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: I'll definitely make an ask to the torch team to figure it out, and post here. Probably before we merge this I imagine?

To the point around giant memory leak, I'm not sure that'll be the case to be honest / I'm not particularly worried. These methods work as is if you disable torch compile, so that tells me that the closures are already somewhere. When profiling the script above with memray, I see that we have allocations / memory use within a normal wiggle room (+-5 allocs, 50kb). Here are the zips.
[prof.zip](https://github.com/user-attachments/files/17822043/prof.zip)

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/basic.py:66]: Isn't this setting am ever increasing number of lambdas as an attribute of that module? 

We once had a problem because we were creating dispatch functions for dynamic Ops in PyMC and someone complained about it (there will always be someone): https://github.com/pymc-devs/pymc/issues/6990

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: Ah, I can explore around with the geographic- but iiuc it just stores another reference to the lambda, it shouldn't be an explicit copy. I could verify by say doing a deep copy and we should see the number go up.

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/basic.py:66]: But the point is that lambda will never be gc while PyTensor is around no? Even after the torch function disappears?

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: Yup, that's right.

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: I guess it should get gc'd when `link.utils` goes out of scope...unsure when that is

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: This is the issue we end up seeing

```txt
torch._dynamo.exc.InternalTorchDynamoError: weakly-referenced object no longer exists

from user code:
   File "/var/folders/98/g1t2_d2x4w94vqfv06xhyz6c0000gp/T/tmpnk75e974", line 3, in pytorch_funcified_fgraph
    tensor_variable_2, tensor_variable_3 = pytorch_funcified_fgraph(y, z)
```

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/basic.py:66]: Is there a way to make the functiongraph that torch does keep around to be the one referening those closures? The link.utils will never go away unless someone does `del pytensor`

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: Maybe we can a `__del__`method in the torch linker, and have that cleaned up these explicit references? 

Additionally yea I'll see if I cheese something where we have the global fgraph hold a reference somewhere....

Comment (Ch0ronomato) [reply pytensor/link/pytorch/dispatch/basic.py:66]: I tried to add a sort of wrapper execution function that can just allow us to clean up after ourselves. I do make an assumption that each `pytensor.function` def will have it's own `PytorchLinker`, lmk if that is not correct. This successfully cleaned up the references after the the new function goes out of scope.

File: pytensor/link/pytorch/linker.py [hunk 5]
```diff
@@ -18,14 +24,45 @@ def output_filter(self, var: Variable, out: Any) -> Any:
...
         import torch
 
-        return torch.compile(fn)
+        class wrapper:
+            def __init__(self, fn, gen_functors):
+                self.fn = torch.compile(fn)
+                self.gen_functors = copy.copy(gen_functors)
+
+            def __call__(self, *args, **kwargs):
+                import pytensor.link.utils
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/linker.py:45]: Note, there is no way this is threadsafe.

File: pytensor/link/pytorch/dispatch/basic.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/basic.py]: Yes the function will have it's own PyTorchLinker

File: pytensor/link/pytorch/dispatch/basic.py [hunk 6]
```diff
@@ -56,9 +56,17 @@ def pytorch_funcify_FunctionGraph(
     fgraph_name="pytorch_funcified_fgraph",
     **kwargs,
 ):
+    # Ensure that torch is aware of the leaf nodes
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:59]: What leaf nodes? Comment is unclear to me

File: pytensor/link/pytorch/linker.py [hunk 7]
```diff
@@ -18,14 +24,45 @@ def output_filter(self, var: Variable, out: Any) -> Any:
...
         return pytorch_funcify(
-            fgraph, input_storage=input_storage, storage_map=storage_map, **kwargs
+            fgraph, input_storage=input_storage, storage_map=storage_map, **built_kwargs
         )
 
     def jit_compile(self, fn):
         import torch
 
-        return torch.compile(fn)
+        class wrapper:
```
Comment (ricardoV94) [inline pytensor/link/pytorch/linker.py:39]: add some docstrings explaining why we're doing this and that we can hopefully revert in the future? Link to issue on torch if you opened one already

File: pytensor/link/pytorch/linker.py [hunk 8]
```diff
@@ -34,3 +71,6 @@ def create_thunk_inputs(self, storage_map):
             thunk_inputs.append(sinput)
 
         return thunk_inputs
+
+    def record_fn(self, name, fn):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/linker.py:75]: Is this used?

File: pytensor/link/utils.py [hunk 9]
```diff
@@ -704,6 +704,8 @@ def fgraph_to_python(
     get_name_for_object
         A function used to provide names for the objects referenced within the
         generated function.
+    unique_name
```
Comment (ricardoV94) [inline pytensor/link/utils.py:707]: Don't pass through kwargs here but explicitly with the default being what we had before

**Review by ricardoV94 — COMMENTED** [review]
Looks better. Can you add a test showing the extra attributes on the module being discarded when you del the function?

Also can we store them with a leading _? Nit to clobber up during debugging?

File: pytensor/link/pytorch/linker.py (hunk 8 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/linker.py:75]: Yup, here https://github.com/pymc-devs/pytensor/pull/1094/files#diff-bc4aa470accf91cb330a5f1f6949b9599d4da6c341290e78e27d684144b41429R64

Comment (Ch0ronomato) [reply pytensor/link/pytorch/linker.py]: I actually removed this and moved the "conversion" func into the linker, so now the linker is doing all the tracking and stuff. Seemed more reasonable.

File: tests/link/pytorch/test_blockwise.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
 basic = pytest.importorskip("pytensor.link.pytorch.dispatch.basic")
 
 
-class TestOp(Op):
+class BatchedTestOp(Op):
```
Comment (Ch0ronomato) [inline tests/link/pytorch/test_blockwise.py:15]: We can reverse this, i was just getting annoyed with the pytest warning

Comment (ricardoV94) [reply tests/link/pytorch/test_blockwise.py:15]: Makes sense, just make sure to be in a separate commit

**Comment by Ch0ronomato** [comment]
Okay @ricardoV94 if this still looks good, it should be ready to merge.

**Review by ricardoV94 — APPROVED** [review]
